<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8">
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />

<meta name="viewport" content="width=device-width, initial-scale=1">

<meta name="author" content="Kaur Alasoo" />

<meta name="date" content="2017-01-08" />

<title>Introduction to wiggleplotr</title>



<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
</style>



<link href="data:text/css;charset=utf-8,body%20%7B%0Abackground%2Dcolor%3A%20%23fff%3B%0Amargin%3A%201em%20auto%3B%0Amax%2Dwidth%3A%20700px%3B%0Aoverflow%3A%20visible%3B%0Apadding%2Dleft%3A%202em%3B%0Apadding%2Dright%3A%202em%3B%0Afont%2Dfamily%3A%20%22Open%20Sans%22%2C%20%22Helvetica%20Neue%22%2C%20Helvetica%2C%20Arial%2C%20sans%2Dserif%3B%0Afont%2Dsize%3A%2014px%3B%0Aline%2Dheight%3A%201%2E35%3B%0A%7D%0A%23header%20%7B%0Atext%2Dalign%3A%20center%3B%0A%7D%0A%23TOC%20%7B%0Aclear%3A%20both%3B%0Amargin%3A%200%200%2010px%2010px%3B%0Apadding%3A%204px%3B%0Awidth%3A%20400px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Aborder%2Dradius%3A%205px%3B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Afont%2Dsize%3A%2013px%3B%0Aline%2Dheight%3A%201%2E3%3B%0A%7D%0A%23TOC%20%2Etoctitle%20%7B%0Afont%2Dweight%3A%20bold%3B%0Afont%2Dsize%3A%2015px%3B%0Amargin%2Dleft%3A%205px%3B%0A%7D%0A%23TOC%20ul%20%7B%0Apadding%2Dleft%3A%2040px%3B%0Amargin%2Dleft%3A%20%2D1%2E5em%3B%0Amargin%2Dtop%3A%205px%3B%0Amargin%2Dbottom%3A%205px%3B%0A%7D%0A%23TOC%20ul%20ul%20%7B%0Amargin%2Dleft%3A%20%2D2em%3B%0A%7D%0A%23TOC%20li%20%7B%0Aline%2Dheight%3A%2016px%3B%0A%7D%0Atable%20%7B%0Amargin%3A%201em%20auto%3B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dcolor%3A%20%23DDDDDD%3B%0Aborder%2Dstyle%3A%20outset%3B%0Aborder%2Dcollapse%3A%20collapse%3B%0A%7D%0Atable%20th%20%7B%0Aborder%2Dwidth%3A%202px%3B%0Apadding%3A%205px%3B%0Aborder%2Dstyle%3A%20inset%3B%0A%7D%0Atable%20td%20%7B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dstyle%3A%20inset%3B%0Aline%2Dheight%3A%2018px%3B%0Apadding%3A%205px%205px%3B%0A%7D%0Atable%2C%20table%20th%2C%20table%20td%20%7B%0Aborder%2Dleft%2Dstyle%3A%20none%3B%0Aborder%2Dright%2Dstyle%3A%20none%3B%0A%7D%0Atable%20thead%2C%20table%20tr%2Eeven%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Ap%20%7B%0Amargin%3A%200%2E5em%200%3B%0A%7D%0Ablockquote%20%7B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Apadding%3A%200%2E25em%200%2E75em%3B%0A%7D%0Ahr%20%7B%0Aborder%2Dstyle%3A%20solid%3B%0Aborder%3A%20none%3B%0Aborder%2Dtop%3A%201px%20solid%20%23777%3B%0Amargin%3A%2028px%200%3B%0A%7D%0Adl%20%7B%0Amargin%2Dleft%3A%200%3B%0A%7D%0Adl%20dd%20%7B%0Amargin%2Dbottom%3A%2013px%3B%0Amargin%2Dleft%3A%2013px%3B%0A%7D%0Adl%20dt%20%7B%0Afont%2Dweight%3A%20bold%3B%0A%7D%0Aul%20%7B%0Amargin%2Dtop%3A%200%3B%0A%7D%0Aul%20li%20%7B%0Alist%2Dstyle%3A%20circle%20outside%3B%0A%7D%0Aul%20ul%20%7B%0Amargin%2Dbottom%3A%200%3B%0A%7D%0Apre%2C%20code%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0Aborder%2Dradius%3A%203px%3B%0Acolor%3A%20%23333%3B%0Awhite%2Dspace%3A%20pre%2Dwrap%3B%20%0A%7D%0Apre%20%7B%0Aborder%2Dradius%3A%203px%3B%0Amargin%3A%205px%200px%2010px%200px%3B%0Apadding%3A%2010px%3B%0A%7D%0Apre%3Anot%28%5Bclass%5D%29%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Acode%20%7B%0Afont%2Dfamily%3A%20Consolas%2C%20Monaco%2C%20%27Courier%20New%27%2C%20monospace%3B%0Afont%2Dsize%3A%2085%25%3B%0A%7D%0Ap%20%3E%20code%2C%20li%20%3E%20code%20%7B%0Apadding%3A%202px%200px%3B%0A%7D%0Adiv%2Efigure%20%7B%0Atext%2Dalign%3A%20center%3B%0A%7D%0Aimg%20%7B%0Abackground%2Dcolor%3A%20%23FFFFFF%3B%0Apadding%3A%202px%3B%0Aborder%3A%201px%20solid%20%23DDDDDD%3B%0Aborder%2Dradius%3A%203px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Amargin%3A%200%205px%3B%0A%7D%0Ah1%20%7B%0Amargin%2Dtop%3A%200%3B%0Afont%2Dsize%3A%2035px%3B%0Aline%2Dheight%3A%2040px%3B%0A%7D%0Ah2%20%7B%0Aborder%2Dbottom%3A%204px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Apadding%2Dbottom%3A%202px%3B%0Afont%2Dsize%3A%20145%25%3B%0A%7D%0Ah3%20%7B%0Aborder%2Dbottom%3A%202px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Afont%2Dsize%3A%20120%25%3B%0A%7D%0Ah4%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23f7f7f7%3B%0Amargin%2Dleft%3A%208px%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Ah5%2C%20h6%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23ccc%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Aa%20%7B%0Acolor%3A%20%230033dd%3B%0Atext%2Ddecoration%3A%20none%3B%0A%7D%0Aa%3Ahover%20%7B%0Acolor%3A%20%236666ff%3B%20%7D%0Aa%3Avisited%20%7B%0Acolor%3A%20%23800080%3B%20%7D%0Aa%3Avisited%3Ahover%20%7B%0Acolor%3A%20%23BB00BB%3B%20%7D%0Aa%5Bhref%5E%3D%22http%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0Aa%5Bhref%5E%3D%22https%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0A%0Acode%20%3E%20span%2Ekw%20%7B%20color%3A%20%23555%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Edt%20%7B%20color%3A%20%23902000%3B%20%7D%20%0Acode%20%3E%20span%2Edv%20%7B%20color%3A%20%2340a070%3B%20%7D%20%0Acode%20%3E%20span%2Ebn%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Efl%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Ech%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Est%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Eco%20%7B%20color%3A%20%23888888%3B%20font%2Dstyle%3A%20italic%3B%20%7D%20%0Acode%20%3E%20span%2Eot%20%7B%20color%3A%20%23007020%3B%20%7D%20%0Acode%20%3E%20span%2Eal%20%7B%20color%3A%20%23ff0000%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Efu%20%7B%20color%3A%20%23900%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%20code%20%3E%20span%2Eer%20%7B%20color%3A%20%23a61717%3B%20background%2Dcolor%3A%20%23e3d2d2%3B%20%7D%20%0A" rel="stylesheet" type="text/css" />

</head>

<body>




<h1 class="title toc-ignore">Introduction to wiggleplotr</h1>
<h4 class="author"><em>Kaur Alasoo</em></h4>
<h4 class="date"><em>2017-01-08</em></h4>



<div id="using-biomart-to-download-transcript-annotations-from-ensembl" class="section level2">
<h2>Using biomaRt to download transcript annotations from Ensembl</h2>
<p>The easiest way to access reference transcript annotations in R is to download them directly from Ensembl using the biomaRt R package. To do that, we first need to load the neccessary R packages.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(<span class="st">&quot;biomaRt&quot;</span>)
<span class="kw">library</span>(<span class="st">&quot;dplyr&quot;</span>)
<span class="kw">library</span>(<span class="st">&quot;GenomicFeatures&quot;</span>)</code></pre></div>
<div id="downloading-transcript-metadata" class="section level3">
<h3>Downloading transcript metadata</h3>
<p>First, we want to download transcript metadata, such as which transcripts belong to which genes and what are their names. We can use the biomaRt package to do that. First, let’s define which mart and dataset we want to use.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">ensembl_mart =<span class="st"> </span><span class="kw">useMart</span>(<span class="st">&quot;ENSEMBL_MART_ENSEMBL&quot;</span>, <span class="dt">host =</span> <span class="st">&quot;dec2014.archive.ensembl.org&quot;</span>)
ensembl_dataset =<span class="st"> </span><span class="kw">useDataset</span>(<span class="st">&quot;hsapiens_gene_ensembl&quot;</span>,<span class="dt">mart=</span>ensembl_mart)
ensembl_dataset</code></pre></div>
<pre><code>## Object of class 'Mart':
##  Using the ENSEMBL_MART_ENSEMBL BioMart database
##  Using the hsapiens_gene_ensembl dataset</code></pre>
<p>The <code>host</code> helps to make sure that we get the annotations from a specific Ensembl version. For example, Ensembl 78 correseponds to <code>host=&quot;dec2014.archive.ensembl.org&quot;</code>. You can use the Ensembl Archives <a href="http://www.ensembl.org/info/website/archives/index.html">website</a> to check which host name corresponds to desired Ensembl version. More information using specific ensembl versions with biomaRt can be found in the <a href="https://www.bioconductor.org/packages/3.3/bioc/vignettes/biomaRt/inst/doc/biomaRt.pdf">biomaRt vignette</a>.</p>
<p>We can see all available attributes with the <code>listAttributes</code> command.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">attributes =<span class="st"> </span><span class="kw">listAttributes</span>(ensembl_dataset)
<span class="kw">head</span>(attributes)</code></pre></div>
<pre><code>##                    name           description         page
## 1       ensembl_gene_id       Ensembl Gene ID feature_page
## 2 ensembl_transcript_id Ensembl Transcript ID feature_page
## 3    ensembl_peptide_id    Ensembl Protein ID feature_page
## 4       ensembl_exon_id       Ensembl Exon ID feature_page
## 5           description           Description feature_page
## 6       chromosome_name       Chromosome Name feature_page</code></pre>
<p>Now, let’s select gene id, gene name, transcript_id and strand from the biomart and download the corresponding columns.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">selected_attributes =<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;ensembl_transcript_id&quot;</span>, <span class="st">&quot;ensembl_gene_id&quot;</span>, <span class="st">&quot;external_gene_name&quot;</span>, <span class="st">&quot;strand&quot;</span>,<span class="st">&quot;gene_biotype&quot;</span>, <span class="st">&quot;transcript_biotype&quot;</span>)
data =<span class="st"> </span><span class="kw">getBM</span>(<span class="dt">attributes =</span> selected_attributes, <span class="dt">mart =</span> ensembl_dataset)
<span class="kw">head</span>(data)</code></pre></div>
<pre><code>##   ensembl_transcript_id ensembl_gene_id external_gene_name strand
## 1       ENST00000508957 ENSG00000197468      RP11-747H12.1      1
## 2       ENST00000435337 ENSG00000231049            OR52B5P      1
## 3       ENST00000618935 ENSG00000276385       RP5-859I17.3     -1
## 4       ENST00000614589 ENSG00000275151       RP11-42L13.2      1
## 5       ENST00000432676 ENSG00000228913                UBD     -1
## 6       ENST00000449391 ENSG00000231948         HS1BP3-IT1     -1
##             gene_biotype     transcript_biotype
## 1 unprocessed_pseudogene unprocessed_pseudogene
## 2 unprocessed_pseudogene unprocessed_pseudogene
## 3   processed_pseudogene   processed_pseudogene
## 4   processed_pseudogene   processed_pseudogene
## 5         protein_coding         protein_coding
## 6         sense_intronic         sense_intronic</code></pre>
<p>Finally, we need to rename the columns</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">data =<span class="st"> </span>dplyr::<span class="kw">rename</span>(data, <span class="dt">transcript_id =</span> ensembl_transcript_id, <span class="dt">gene_id =</span> ensembl_gene_id, <span class="dt">gene_name =</span> external_gene_name)
<span class="kw">head</span>(data)</code></pre></div>
<pre><code>##     transcript_id         gene_id     gene_name strand
## 1 ENST00000508957 ENSG00000197468 RP11-747H12.1      1
## 2 ENST00000435337 ENSG00000231049       OR52B5P      1
## 3 ENST00000618935 ENSG00000276385  RP5-859I17.3     -1
## 4 ENST00000614589 ENSG00000275151  RP11-42L13.2      1
## 5 ENST00000432676 ENSG00000228913           UBD     -1
## 6 ENST00000449391 ENSG00000231948    HS1BP3-IT1     -1
##             gene_biotype     transcript_biotype
## 1 unprocessed_pseudogene unprocessed_pseudogene
## 2 unprocessed_pseudogene unprocessed_pseudogene
## 3   processed_pseudogene   processed_pseudogene
## 4   processed_pseudogene   processed_pseudogene
## 5         protein_coding         protein_coding
## 6         sense_intronic         sense_intronic</code></pre>
<p>We can now save the metadata into a file to avoid downloading it every time we need to use it.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">saveRDS</span>(data, <span class="st">&quot;transcript_metadata.rds&quot;</span>)</code></pre></div>
<p>Next time that we need to access the data we can load it directly from disk.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">transcript_metadata =<span class="st"> </span><span class="kw">readRDS</span>(<span class="st">&quot;transcript_metadata.rds&quot;</span>)
<span class="kw">head</span>(transcript_metadata)</code></pre></div>
<pre><code>##     transcript_id         gene_id     gene_name strand
## 1 ENST00000508957 ENSG00000197468 RP11-747H12.1      1
## 2 ENST00000435337 ENSG00000231049       OR52B5P      1
## 3 ENST00000618935 ENSG00000276385  RP5-859I17.3     -1
## 4 ENST00000614589 ENSG00000275151  RP11-42L13.2      1
## 5 ENST00000432676 ENSG00000228913           UBD     -1
## 6 ENST00000449391 ENSG00000231948    HS1BP3-IT1     -1
##             gene_biotype     transcript_biotype
## 1 unprocessed_pseudogene unprocessed_pseudogene
## 2 unprocessed_pseudogene unprocessed_pseudogene
## 3   processed_pseudogene   processed_pseudogene
## 4   processed_pseudogene   processed_pseudogene
## 5         protein_coding         protein_coding
## 6         sense_intronic         sense_intronic</code></pre>
</div>
<div id="downloading-the-full-transcript-database-from-ensembl" class="section level3">
<h3>Downloading the full transcript database from Ensembl</h3>
<p>However, just the transcript metadata is not enought to use wiggleplotr, we also need the coordinates for all exons. We can get those using the <em>GenomicFeatures</em> packages. First, we use the <code>makeTranscriptDbFromBiomart</code> function to download the full transcript database correspondint to a sepcifc Ensembl version, in this case Ensembl 78. Please note that as the database is quite big this can take at least a couple of minutes to run.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">txdb =<span class="st"> </span><span class="kw">makeTranscriptDbFromBiomart</span>(<span class="dt">biomart =</span> <span class="st">&quot;ENSEMBL_MART_ENSEMBL&quot;</span>, <span class="dt">dataset =</span> <span class="st">&quot;hsapiens_gene_ensembl&quot;</span>, <span class="dt">host=</span><span class="st">&quot;dec2014.archive.ensembl.org&quot;</span>)</code></pre></div>
<p>We can also save the database to disk to avoid re-downloading it every time we want to use it.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">saveDb</span>(txdb, <span class="st">&quot;TranscriptDb_GRCh38_78.db&quot;</span>)</code></pre></div>
<p>And we can load it from disk using the loadDb function.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">txdb =<span class="st"> </span><span class="kw">loadDb</span>(<span class="st">&quot;TranscriptDb_GRCh38_79.db&quot;</span>)</code></pre></div>
<p>Finally, we can extract exon and CDS coordinates for all annotated transcripts from the database. This command will produce a a list of [GRanges] objects, each one containing the exons of a single transcript.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">exons =<span class="st"> </span><span class="kw">exonsBy</span>(txdb, <span class="dt">by =</span> <span class="st">&quot;tx&quot;</span>, <span class="dt">use.names =</span> <span class="ot">TRUE</span>)
cdss =<span class="st"> </span><span class="kw">cdsBy</span>(txdb, <span class="dt">by =</span> <span class="st">&quot;tx&quot;</span>, <span class="dt">use.names =</span> <span class="ot">TRUE</span>)
exons[[<span class="st">&quot;ENST00000392477&quot;</span>]]</code></pre></div>
<pre><code>## GRanges object with 16 ranges and 3 metadata columns:
##        seqnames                 ranges strand |   exon_id       exon_name
##           &lt;Rle&gt;              &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt;     &lt;character&gt;
##    [1]        6 [125790763, 125791067]      + |    272039 ENSE00001617948
##    [2]        6 [125815291, 125815404]      + |    272046 ENSE00003634530
##    [3]        6 [125855020, 125855240]      + |    272048 ENSE00003586300
##    [4]        6 [125874889, 125874968]      + |    272052 ENSE00003504190
##    [5]        6 [125878263, 125878370]      + |    272055 ENSE00002453865
##    ...      ...                    ...    ... .       ...             ...
##   [12]        6 [125920943, 125921068]      + |    272068 ENSE00003754534
##   [13]        6 [125922682, 125922834]      + |    272070 ENSE00003730044
##   [14]        6 [125927663, 125927758]      + |    272072 ENSE00003645935
##   [15]        6 [125928174, 125928247]      + |    272073 ENSE00002494782
##   [16]        6 [125928636, 125932030]      + |    272077 ENSE00002498964
##        exon_rank
##        &lt;integer&gt;
##    [1]         1
##    [2]         2
##    [3]         3
##    [4]         4
##    [5]         5
##    ...       ...
##   [12]        12
##   [13]        13
##   [14]        14
##   [15]        15
##   [16]        16
##   -------
##   seqinfo: 869 sequences from an unspecified genome; no seqlengths</code></pre>
</div>
</div>



<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
